### PR TITLE
Speed up Entity List properties fetching by adding Entities.getMultipleEntities API

### PR DESCRIPTION
**Pull request by thoys**
## QA tests

1. Run the following test case: https://highfidelity.testrail.net/index.php?/tests/view/120401

2. Run the Client Script tests
   1. Enable Developer menu (Settings -> Developer Menu)
   2. Go to your sandbox or serverless-domain
   3. Run `Developer -> Tests -> Run Client Script Tests` ( a dialog appears)
   4. All the Entity tests should pass, you can ignore all the other unit tests for this QA. (if you are on serverless it is also expected that `Entity serversExist` will fail)
image

(OSX only update,  apparently Run Client Script Tests doesn't seem to work on the OSX,   open `Developer -> Scripting Console` instead and paste `Script.include("https://raw.githubusercontent.com/highfidelity/hifi/eeda7400c99b1c61472e06d9642ae93b1878e361/scripts/developer/tests/unit_tests/testRunner.js");` )

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13146/HighFidelity-Beta-PR14042-d1f35c8.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13150/HighFidelity-Beta-PR14042-1094b67.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13150/HighFidelity-Beta-PR14042-1094b67.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13150/HighFidelity-Beta-Interface-PR14042-1094b67.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13150/HighFidelity-Beta-PR14042-1094b67.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13150/HighFidelity-Beta-Interface-PR14042-1094b67.exe

File: libraries/entities/src/EntityPsuedoPropertyFlags.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    created = 1 << 3,
+    age = 1 << 4,
+    ageAsText = 1 << 5,
+    lastEdited = 1 << 6,
+    boundingBox = 1 << 7,
+    originalTextures = 1 << 8,
+    renderInfo = 1 << 9,
+    clientOnly = 1 << 10,
+    owningAvatarID = 1 << 11,
+    all = (1 << 12) - 1
```
Comment (huffman) [inline libraries/entities/src/EntityPsuedoPropertyFlags.h]: Can we use all caps here?

File: libraries/entities/src/EntityItemProperties.cpp [hunk 2]
```diff
@@ -1216,34 +1216,49 @@ EntityPropertyFlags EntityItemProperties::getChangedProperties() const {
  * });
  */
 
-QScriptValue EntityItemProperties::copyToScriptValue(QScriptEngine* engine, bool skipDefaults, bool allowUnknownCreateTime, bool strictSemantics) const {
+QScriptValue EntityItemProperties::copyToScriptValue(QScriptEngine* engine, bool skipDefaults, bool allowUnknownCreateTime,
+    bool strictSemantics, EntityPsuedoPropertyFlags psueudoPropertyFlags) const {
```
Comment (huffman) [inline libraries/entities/src/EntityItemProperties.cpp]: Thought on using `PropertyFlags<EntityPseudoPropertyList>` or `std::bitset<EntityPseudoPropertyList>` instead?

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 3]
```diff
@@ -399,6 +399,136 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
+    return entityScriptingInterface->getMultipleEntityPropertiesInternal(engine, entityIDs, context->argument(ARGUMENT_EXTENDED_DESIRED_PROPERTIES));
+}
+
+QScriptValue EntityScriptingInterface::getMultipleEntityPropertiesInternal(QScriptEngine* engine, QVector<QUuid> entityIDs, const QScriptValue& extendedDesiredProperties) {
+    PROFILE_RANGE(script_entities, __FUNCTION__);
+
+    EntityPsuedoPropertyFlags psuedoPropertyFlags;
+    const auto readExtendedPropertyStringValue = [&](QScriptValue extendedProperty) {
+        const auto extendedPropertyString = extendedProperty.toString();
+        if (extendedPropertyString == QString("id")) {
```
Comment (huffman) [inline libraries/entities/src/EntityScriptingInterface.cpp]: I don't think the `QString(...)` invocation is required here, is it?

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 4]
```diff
@@ -399,6 +399,136 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
+    } else if (extendedDesiredProperties.isArray()) {
+        const quint32 length = extendedDesiredProperties.property("length").toInt32();
+        for (quint32 i = 0; i < length; i++) {
+            readExtendedPropertyStringValue(extendedDesiredProperties.property(i));
+        }
+        psuedoPropertyFlags |= EntityPsuedoPropertyFlag::flagsActive;
+    }
+
+    EntityPropertyFlags desiredProperties = qScriptValueToValue<EntityPropertyFlags>(extendedDesiredProperties);
+    bool needsScriptSymantics = desiredProperties.getHasProperty(PROP_POSITION) ||
```
Comment (huffman) [inline libraries/entities/src/EntityScriptingInterface.cpp]: This should be `Semantics`

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 5]
```diff
@@ -399,6 +399,136 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
+                        EntityPropertiesResult result(properties, entity->getScalesWithParent());
+                        resultProperties.append(result);
+                    }
+                }
+            });
+        }
+    }
+    QScriptValue finalResult = engine->newArray(resultProperties.size());
+    quint32 i = 0;
+    {
```
Comment (huffman) [inline libraries/entities/src/EntityScriptingInterface.cpp]: is this block still needed?

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 6]
```diff
@@ -399,6 +399,136 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
+        if (needsScriptSymantics) {
+            PROFILE_RANGE(script_entities, "EntityScriptingInterface::getMultipleEntityProperties>Script Semantics");
+            foreach(auto result, resultProperties) {
```
Comment (huffman) [inline libraries/entities/src/EntityScriptingInterface.cpp]: can we use `for (const auto& result : resultProperties)` here instead?

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 7]
```diff
@@ -399,6 +399,136 @@ EntityItemProperties EntityScriptingInterface::getEntityProperties(QUuid identit
...
+                finalResult.setProperty(i++, convertPropertiesToScriptSemantics(result.properties, result.scalesWithParent)
+                    .copyToScriptValue(engine, false, false, false, psuedoPropertyFlags));
+            }
+        } else {
+            PROFILE_RANGE(script_entities, "EntityScriptingInterface::getMultipleEntityProperties>Skip Script Semantics");
+            foreach(auto result, resultProperties) {
```
Comment (huffman) [inline libraries/entities/src/EntityScriptingInterface.cpp]: Same comment as above

File: libraries/entities/src/EntityItemProperties.cpp (hunk 2 above)
Comment (thoys) [reply libraries/entities/src/EntityItemProperties.cpp]: ✔️

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13325/HighFidelity-Beta-PR14042-80beb6c.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13325/HighFidelity-Beta-PR14042-80beb6c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13325/HighFidelity-Beta-Interface-PR14042-80beb6c.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13479/HighFidelity-Beta-PR14042-e266e2e.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13479/HighFidelity-Beta-PR14042-e266e2e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13479/HighFidelity-Beta-Interface-PR14042-e266e2e.exe

**Comment by QliCknox** [comment]
Pass @conklin94122 

Users were able to run the script and get expected results. Entity test case passed

**Comment by huffman** [comment]
@JeffClinton I believe we can just rerun the unit tests. Those are the only things that will change.

**Comment by QliCknox** [comment]
Currently Testing

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13506/HighFidelity-Beta-PR14042-eeda740.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13506/HighFidelity-Beta-Interface-PR14042-eeda740.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13506/HighFidelity-Beta-PR14042-eeda740.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13506/HighFidelity-Beta-Interface-PR14042-eeda740.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13511/HighFidelity-Beta-PR14042-eeda740.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13511/HighFidelity-Beta-PR14042-eeda740.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13511/HighFidelity-Beta-Interface-PR14042-eeda740.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13511/HighFidelity-Beta-PR14042-eeda740.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13511/HighFidelity-Beta-Interface-PR14042-eeda740.dmg

**Pull request merged**